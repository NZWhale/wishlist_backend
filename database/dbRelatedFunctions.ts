import {IAuthRequestRow, ISessionRow, IUserRow, IWishRow, IWishListDb, IRoomRow} from "./interfaces";
import {roomIdLength, wishIdLength} from "../addresses";
import createRandomId from "../createRandomId";

type Email = string
type UserId = string
type WishId = string
type Cookie = string
type MagicId = string
type RoomId = string
type Table = 'authRequests' | 'sessions'

export const createAuthRequestRecord = (dbContent: IWishListDb, magicId: MagicId, userEmail: Email) => {
    const authRequestRecord: IAuthRequestRow = {
        token: magicId,
        email: userEmail
    }
    dbContent.authRequests.push(authRequestRecord)
}

export const createUserRecord = (dbContent: IWishListDb, userId: UserId, userEmail: Email) => {
    const userRecord: IUserRow = {
        userId: userId,
        email: userEmail,
        username: null
    }
    dbContent.users.push(userRecord)
}

export const createSessionRecord = (dbContent: IWishListDb, userId: UserId, cookie: Cookie) => {
    const sessionRecord: ISessionRow = {
        userId: userId,
        cookie: cookie
    }
    dbContent.sessions.push(sessionRecord)
}

export const isAuthRequestExist = (dbContent: IWishListDb, email: Email) => {
    const result = dbContent.authRequests.findIndex((authRequest: IAuthRequestRow) => authRequest.email === email)
    return result !== -1
}
export const isAuthRequestExistByToken = (dbContent: IWishListDb, token: string) => {
    const result = dbContent.authRequests.findIndex((authRequest: IAuthRequestRow) => authRequest.token === token)
    return result !== -1
}

export const returnWishIndex = (dbContent: IWishListDb, wishId: WishId) => {
    const result = dbContent.wishes.findIndex((wishRow: IWishRow) => wishRow.wishId === wishId)
    if (result === -1) {
        return false
    }
    return result
}

export const isUserExist = (dbContent: IWishListDb, email: Email) => {
    const result = dbContent.users.findIndex(user => user.email === email)
    return result !== -1
}

export const isSessionExist = (dbContent: IWishListDb, userId: UserId) => {
    const result = dbContent.sessions.findIndex((session: ISessionRow) => session.userId === userId)
    return result !== -1
}

export const getUserEmailFromAuthRequests = (dbContent: IWishListDb, token: string) => {
    const result = dbContent.authRequests.find((authRequest: IAuthRequestRow) => authRequest.token === token)
    return result ? result.email : ""
}

export const getUserIdFromDb = (dbContent: IWishListDb, email: Email) => {
    const result = dbContent.users.find((user: IUserRow) => user.email === email)
    return result ? result.userId : ""
}

export const getUserIdByCookie = (dbContent: IWishListDb, cookie: Cookie) => {
    const result = dbContent.sessions.find((session: ISessionRow) => session.cookie === cookie)
    return result ? result.userId : ""
}

export const getUsernameByUserId = (dbContent: IWishListDb, userId: UserId): string | null => {
    const username = dbContent.users.find((user: IUserRow) => user.userId === userId)
    return username ? username.username : null
}

export const getUserIdByEmail = (dbContent: IWishListDb, email: string) => {
    const result = dbContent.users.find((user: IUserRow) => user.email === email)
    return result ? result.userId : null
}
export const getUserIdByUsername = (dbContent: IWishListDb, username: string) => {
    const result = dbContent.users.find((user: IUserRow) => user.username === username)
    return result ? result.userId : null
}


export const getUserData = (dbContent: IWishListDb, email: Email) => {
    return dbContent.users.find((user: IUserRow) => user.email === email)
}

export const getAllWishesOfLoggedInUser = (dbContent: IWishListDb, userId: UserId): IWishRow[] => {
    const allWishesOfUser: IWishRow[] = []
    dbContent.wishes.forEach((wish: IWishRow) => {
        if (wish.userId === userId) {
            allWishesOfUser.push(wish)
        }
    })
    return allWishesOfUser
}

export const getPublicWishesByUserId = (dbContent: IWishListDb, userId: string): IWishRow[] => {
    const publicWishes: IWishRow[] = []
    dbContent.wishes.forEach((wish: IWishRow) => {
        if (wish.userId === userId && wish.isPublic) {
            publicWishes.push(wish)
        }
    })
    return publicWishes
}

export const deleteContentFromDb = (dbContent: IWishListDb, table: Table, email: Email) => {
    //TODO: create separate functions for each deleting content
    switch (table) {
        case 'authRequests':
            const authRequestIndex = dbContent.authRequests.findIndex((authRequest: IAuthRequestRow) => authRequest.email === email)
            dbContent.authRequests.splice(authRequestIndex, 1)
            break
        case 'sessions':
            const userId = getUserIdFromDb(dbContent, email)
            const sessionIndex = dbContent.sessions.findIndex((session: ISessionRow) => session.userId === userId)
            dbContent.sessions.splice(sessionIndex, 1)
            break
        default:
            return "Incorrect table"
    }

}

export const createNewWishRecord = (dbContent: IWishListDb, userId: UserId, title: string, description: string, isPublic: boolean) => {
    const wishRow: IWishRow = {
        userId: userId,
        wishId: createRandomId(wishIdLength, title),
        title: title,
        description: description,
        isPublic: isPublic
    }
    dbContent.wishes.push(wishRow)
}
export const deleteWishRecord = (dbContent: IWishListDb, wishIndex: number) => {
    dbContent.wishes.splice(wishIndex, 1)
}

export const editWishRecord = (dbContent: IWishListDb, wishIndex: number, isPublic: boolean, title?: string, description?: string) => {
    if (title) {
        dbContent.wishes[wishIndex].title = title
    }
    if (description) {
        dbContent.wishes[wishIndex].description = description
    }
    dbContent.wishes[wishIndex].isPublic = isPublic

}

export const setUsername = (dbContent: IWishListDb, userId: UserId, nickname: string) => {
    dbContent.users.forEach((user: IUserRow) => {
        if (user.userId === userId) {
            user.username = nickname
            return
        }
    })
}

export const getUserDataByUserId = (dbContent: IWishListDb, userId: UserId) => {
    const user = dbContent.users.find((user: IUserRow) => user.userId === userId)
    return user ? user : undefined
}

export const getRoomIdByRoomName = (dbContent: IWishListDb, roomName: string) => {
    const room = dbContent.rooms.find((room: IRoomRow) => room.roomName === roomName)
    return room ? room.roomId : undefined
}

export const createNewRoom = (dbContent: IWishListDb, userId: UserId, roomName: string) => {
    const user = getUserDataByUserId(dbContent, userId)
    if (!user) {
        throw new Error("User doesn't exist")
    }
    const roomRow: IRoomRow = {
        creatorId: userId,
        roomId: createRandomId(roomIdLength, roomName),
        roomName: roomName,
        users: []
    }
    roomRow.users.push(user.userId)
    dbContent.rooms.push(roomRow)
}

export const addUserToRoomTable = (dbContent: IWishListDb, roomId: RoomId, addableUserId: UserId) => {
    const user = getUserDataByUserId(dbContent, addableUserId)
    if (!user) {
        throw new Error("User doesn't exist")
    }
    dbContent.rooms.forEach((room: IRoomRow) => {
        if (room.roomId === roomId) {
            room.users.push(user.userId)
        }
    })

}

export const getRoomIdByCreaterId = (dbContent: IWishListDb, roomCreatorId: UserId) => {
    const room = dbContent.rooms.find((room: IRoomRow) => room.creatorId === roomCreatorId)
    return room ? room.roomId : undefined
}

export const getAllRoomsOfUser = (dbContent: IWishListDb, userId: UserId) => {
    const allRoomsOfUser: IRoomRow[] = []
    dbContent.rooms.forEach((room: IRoomRow) => {
        room.users.forEach((user: string) => {
            console.log(user, userId)
            if(user === userId){
                allRoomsOfUser.push(room)
            }
            })
        })
    if(allRoomsOfUser.length === 0) {
        throw new Error("User isn't a member of any rooms")
    }
    return allRoomsOfUser
}

export const createEmptyDbContent = (): IWishListDb => {
    return {
        "rooms": [],
        "users": [],
        "sessions": [],
        "wishes": [],
        "authRequests": []
    }
}