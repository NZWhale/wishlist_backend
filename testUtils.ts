import path from "path";
import tmp from "tmp";
import WishListFileDatabase from "./database/Database";

export const createEmptyTestDatabase = () => {
    const dbFilePath = path.join(tmp.dirSync().name, "db.json")
    return {db: new WishListFileDatabase(dbFilePath), dbFilePath}
}