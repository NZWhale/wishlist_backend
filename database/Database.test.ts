import fs from "fs";
import {getRoomIdByRoomName, setUsername} from "./dbRelatedFunctions";
import {createEmptyTestDatabase} from "../testUtils";

jest.mock('nanoid')
jest.mock("../createRandomId")

describe("createMagicId", () => {
    test("should create and return magic ID", async () => {
        const {db} = createEmptyTestDatabase();
        const magicId = await db.createMagicId("420@chill.com");
        expect(magicId).toBe("x420@chill.comxx")
    })

    test("should create new user entity it doesn't exist", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase();
        await db.createMagicId("420@chill.com");
        const fileContent = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContent).toMatchSnapshot()
    })
})

describe("authoriseUser", () => {
    test("should delete authorise request and create user session", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        const magicId = await db.createMagicId("420@chill.com");
        const fileContentAfterCreatingMagicId = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterCreatingMagicId).toMatchSnapshot()
        const cookie = await db.authoriseUser(magicId)
        expect(cookie).toBe('xx420@chill.comxxx')
        const fileContentAfterAuthoriseUser = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAuthoriseUser).toMatchSnapshot()
    })
})

describe("addNewWish", () => {
    test("should create new wish row in the wishes table", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        const magicId = await db.createMagicId("420@chill.com");
        const fileContentAfterCreatingMagicId = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterCreatingMagicId).toMatchSnapshot()
        const cookie = await db.authoriseUser(magicId)
        await db.addNewWish(cookie, 'Foo', 'Bar', true)
        const fileContentAfterAddingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingWish).toMatchSnapshot()
    })
})

describe("deleteWish", () => {
    test("should delete wish row from wishes table", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        const magicId = await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser(magicId)
        await db.addNewWish(cookie, 'Foo', 'Bar', true)
        const fileContentAfterAddingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingWish).toMatchSnapshot()
        await db.deleteWish('xFooxxxx', cookie)
        const fileContentAfterDeletingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterDeletingWish).toMatchSnapshot()
    })
})

describe("editWish", () => {
    test("should edit wish row", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser("xxxxxxxxxxxxxxxx")
        await db.addNewWish(cookie, 'Foo', 'Bar', true)
        const fileContentAfterAddingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingWish).toMatchSnapshot()
        await db.editWish('xxxxxxxx', 'four', 'twenty', true)
        const fileContentAfterEditingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterEditingWish).toMatchSnapshot()
    })
})

describe("getWishes", () => {
    test("should return all wishes of loggedIn user", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser("xxxxxxxxxxxxxxxx")
        await db.addNewWish(cookie, 'Foo', 'Bar', true)
        await db.addNewWish(cookie, 'Bar', 'Foo', true)
        const fileContentAfterAddingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingWish).toMatchSnapshot()
        const allWishesOfLoggedInUser = await db.getAllWishesOfLoggedInUser(cookie)
        const expectedWishes = [
            {
                "userId": "xxxxxxxxxx",
                "wishId": "xxxxxxxx",
                "title": "Foo",
                "description": "Bar",
                "isPublic": true,
            },
            {
                "userId": "xxxxxxxxxx",
                "wishId": "xxxxxxxx",
                "title": "Bar",
                "description": "Foo",
                "isPublic": true,
            }
        ]
        expect(allWishesOfLoggedInUser).toEqual(expectedWishes)
    })
    test('should return public wishes of user', async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser("xxxxxxxxxxxxxxxx")
        await db.setUsername(cookie, 'FooUser')
        await db.addNewWish(cookie, 'Foo', 'Bar', true)
        await db.addNewWish(cookie, 'Private', 'Foo', false)
        await db.addNewWish(cookie, 'Bar', 'Foo', true)
        await db.addNewWish(cookie, 'Private', 'Bar', false)
        const fileContentAfterAddingWish = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingWish).toMatchSnapshot()
        const publicWishesOfUser = await db.getPublicWishesOfUser('FooUser')
        const expectedWishes = [
            {
                "userId": "xxxxxxxxxx",
                "wishId": "xxxxxxxx",
                "title": "Foo",
                "description": "Bar",
                "isPublic": true,
            },
            {
                "userId": "xxxxxxxxxx",
                "wishId": "xxxxxxxx",
                "title": "Bar",
                "description": "Foo",
                "isPublic": true,
            }
        ]
        expect(publicWishesOfUser).toEqual(expectedWishes)
    })
})

describe("setUsername", () => {
    test("should set username", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser("xxxxxxxxxxxxxxxx")
        const fileContentAfterCreatingUser = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterCreatingUser).toMatchSnapshot()
        await db.setUsername(cookie, 'foobar')
        const fileContentAfterAddingUsername = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingUsername).toMatchSnapshot()
    })
})

describe("getUsername", () => {
    test("should return username", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        const cookie = await db.authoriseUser("xxxxxxxxxxxxxxxx")
        await db.setUsername(cookie, 'foobar')
        const fileContentAfterAddingUsername = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingUsername).toMatchSnapshot()
        const username = await db.getUsernameByCookie(cookie)
        expect(username).toBe('foobar')
    })
})

describe("createRoom", () => {
    test("should create new room in database", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()
        await db.createMagicId("420@chill.com");
        await db.authoriseUser("xxxxxxxxxxxxxxxx")
        const fileContentAfterAuthorise = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAuthorise).toMatchSnapshot()
        await db.createRoom("xxxxxxxxxxxxxxxxxx", "DoobkiRoom")
        const fileContentAfterCreatingRoom = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterCreatingRoom).toMatchSnapshot()
    })
})

describe("addUserToRoom", () => {
    test("should add user to room", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()

        const magicId1 = await db.createMagicId("420@chill.com");
        const cookie1 = await db.authoriseUser(magicId1)
        await db.setUsername(cookie1, 'Vasiliy')

        await db.createRoom(cookie1, "DoobkiRoom")

        const magicId2 = await db.createMagicId("bumpie@gmail.com");
        const cookie2 = await db.authoriseUser(magicId2)
        await db.setUsername(cookie2, 'bumpie')

        const fileContentAfterCreatingRoom = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        const roomId = getRoomIdByRoomName(JSON.parse(fileContentAfterCreatingRoom), "DoobkiRoom")
        if(!roomId) {
            throw new Error("roomId doesn't exist")
        }
        expect(fileContentAfterCreatingRoom).toMatchSnapshot()
        await db.addUserToRoom(cookie1, roomId, "bumpie@gmail.com")
        const fileContentAfterAddingUser = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterAddingUser).toMatchSnapshot()
    })
})

describe("getRoomsOfUser", () => {
    test("should return all rooms of logged in user", async () => {
        const {db, dbFilePath} = createEmptyTestDatabase()

        const magicId1 = await db.createMagicId("420@chill.com");
        const cookie1 = await db.authoriseUser(magicId1)
        await db.setUsername(cookie1, 'foobar')

        await db.createRoom(cookie1, "DoobkiRoom")
        await db.createRoom(cookie1, "Job")
        await db.createRoom(cookie1, "Family")

        const fileContentAfterCreatingRooms = await fs.promises.readFile(dbFilePath, {
            encoding: "utf-8"
        });
        expect(fileContentAfterCreatingRooms).toMatchSnapshot()
        const rooms = await db.getRoomsOfLoggedInUser(cookie1)
        expect(rooms).toMatchSnapshot()
    })
})