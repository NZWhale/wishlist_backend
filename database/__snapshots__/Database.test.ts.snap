// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addNewWish should create new wish row in the wishes table 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [],
  \\"wishes\\": [],
  \\"authRequests\\": [
    {
      \\"token\\": \\"x420@chill.comxx\\",
      \\"email\\": \\"420@chill.com\\"
    }
  ]
}"
`;

exports[`addNewWish should create new wish row in the wishes table 2`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"wishId\\": \\"xFooxxxx\\",
      \\"title\\": \\"Foo\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": true
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`addUserToRoom should add user to room 1`] = `
"{
  \\"rooms\\": [
    {
      \\"creatorId\\": \\"x420@chill\\",
      \\"roomId\\": \\"xDoobkiRoomx\\",
      \\"roomName\\": \\"DoobkiRoom\\",
      \\"users\\": [
        \\"x420@chill\\"
      ]
    }
  ],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": \\"Vasiliy\\"
    },
    {
      \\"userId\\": \\"xbumpie@gm\\",
      \\"email\\": \\"bumpie@gmail.com\\",
      \\"username\\": \\"bumpie\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    },
    {
      \\"userId\\": \\"xbumpie@gm\\",
      \\"cookie\\": \\"xxbumpie@gmail.cox\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`addUserToRoom should add user to room 2`] = `
"{
  \\"rooms\\": [
    {
      \\"creatorId\\": \\"x420@chill\\",
      \\"roomId\\": \\"xDoobkiRoomx\\",
      \\"roomName\\": \\"DoobkiRoom\\",
      \\"users\\": [
        \\"x420@chill\\",
        \\"xbumpie@gm\\"
      ]
    }
  ],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": \\"Vasiliy\\"
    },
    {
      \\"userId\\": \\"xbumpie@gm\\",
      \\"email\\": \\"bumpie@gmail.com\\",
      \\"username\\": \\"bumpie\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    },
    {
      \\"userId\\": \\"xbumpie@gm\\",
      \\"cookie\\": \\"xxbumpie@gmail.cox\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`authoriseUser should delete authorise request and create user session 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [],
  \\"wishes\\": [],
  \\"authRequests\\": [
    {
      \\"token\\": \\"x420@chill.comxx\\",
      \\"email\\": \\"420@chill.com\\"
    }
  ]
}"
`;

exports[`authoriseUser should delete authorise request and create user session 2`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`createMagicId should create new user entity it doesn't exist 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [],
  \\"wishes\\": [],
  \\"authRequests\\": [
    {
      \\"token\\": \\"x420@chill.comxx\\",
      \\"email\\": \\"420@chill.com\\"
    }
  ]
}"
`;

exports[`createRoom should create new room in database 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`createRoom should create new room in database 2`] = `
"{
  \\"rooms\\": [
    {
      \\"creatorId\\": \\"xxxxxxxxxx\\",
      \\"roomId\\": \\"xxxxxxxxxxxx\\",
      \\"roomName\\": \\"DoobkiRoom\\",
      \\"users\\": [
        {
          \\"userId\\": \\"xxxxxxxxxx\\",
          \\"email\\": \\"420@chill.com\\",
          \\"nickname\\": null
        }
      ]
    }
  ],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`deleteWish should delete wish row from wishes table 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"wishId\\": \\"xFooxxxx\\",
      \\"title\\": \\"Foo\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": true
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`deleteWish should delete wish row from wishes table 2`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`editWish should edit wish row 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Foo\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": true
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`editWish should edit wish row 2`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"four\\",
      \\"description\\": \\"twenty\\",
      \\"isPublic\\": true
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`getRoomsOfUser should return all rooms of logged in user 1`] = `
"{
  \\"rooms\\": [
    {
      \\"creatorId\\": \\"x420@chill\\",
      \\"roomId\\": \\"xDoobkiRoomx\\",
      \\"roomName\\": \\"DoobkiRoom\\",
      \\"users\\": [
        \\"x420@chill\\"
      ]
    },
    {
      \\"creatorId\\": \\"x420@chill\\",
      \\"roomId\\": \\"xJobxxxxxxxx\\",
      \\"roomName\\": \\"Job\\",
      \\"users\\": [
        \\"x420@chill\\"
      ]
    },
    {
      \\"creatorId\\": \\"x420@chill\\",
      \\"roomId\\": \\"xFamilyxxxxx\\",
      \\"roomName\\": \\"Family\\",
      \\"users\\": [
        \\"x420@chill\\"
      ]
    }
  ],
  \\"users\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"email\\": \\"420@chill.com\\",
      \\"username\\": \\"foobar\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"x420@chill\\",
      \\"cookie\\": \\"xx420@chill.comxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`getRoomsOfUser should return all rooms of logged in user 2`] = `
Array [
  Object {
    "creatorId": "x420@chill",
    "roomId": "xDoobkiRoomx",
    "roomName": "DoobkiRoom",
    "users": Array [
      "x420@chill",
    ],
  },
  Object {
    "creatorId": "x420@chill",
    "roomId": "xJobxxxxxxxx",
    "roomName": "Job",
    "users": Array [
      "x420@chill",
    ],
  },
  Object {
    "creatorId": "x420@chill",
    "roomId": "xFamilyxxxxx",
    "roomName": "Family",
    "users": Array [
      "x420@chill",
    ],
  },
]
`;

exports[`getUsername should return username 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": \\"foobar\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`getWishes should return all wishes of loggedIn user 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Foo\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": true
    },
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Bar\\",
      \\"description\\": \\"Foo\\",
      \\"isPublic\\": true
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`getWishes should return public wishes of user 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": \\"FooUser\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Foo\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": true
    },
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Private\\",
      \\"description\\": \\"Foo\\",
      \\"isPublic\\": false
    },
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Bar\\",
      \\"description\\": \\"Foo\\",
      \\"isPublic\\": true
    },
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"wishId\\": \\"xxxxxxxx\\",
      \\"title\\": \\"Private\\",
      \\"description\\": \\"Bar\\",
      \\"isPublic\\": false
    }
  ],
  \\"authRequests\\": []
}"
`;

exports[`setUsername should set username 1`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": null
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;

exports[`setUsername should set username 2`] = `
"{
  \\"rooms\\": [],
  \\"users\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"email\\": \\"420@chill.com\\",
      \\"nickname\\": \\"foobar\\"
    }
  ],
  \\"sessions\\": [
    {
      \\"userId\\": \\"xxxxxxxxxx\\",
      \\"cookie\\": \\"xxxxxxxxxxxxxxxxxx\\"
    }
  ],
  \\"wishes\\": [],
  \\"authRequests\\": []
}"
`;
